### ISSUE TITLE
Alarm list sorting question (0.14.7)

### ISSUE THREAD

**Issue by Timo-Fi** [forrestguice/SuntimesWidget#618]
Hello! I have the very same alarm list file imported to Suntimes in both OS7.0 and OS11 devices (I do know that OS11 is not a target device for Suntimes yet, I hope that extra info does not hurt). On both devices it is absolutely not possible to have it so that my 21 alarms would be listed in the order of Alarm Time **in such manner** that the next alarm would be seen topmost on the list.  Interestingly, sorting by Alarm Time does not result the same sort order of the list, between the devices.  The list will always be in order of Alarm Time, but not so that the next alarm would be topmost, there is always some or more alarms on top of the one that will fire next. Both devices have the same (automatic) settings in the OS setup, TZ=+2 and DST=+1.

This alarm list file was initially created using the first version that made it possible to export the list, and there after has been edited on both OS7 and OS11 devices, and using each of the more recent versions of Suntimes. I now updated to 0.14.7, then exported the list from OS11 device, and imported it to OS7 device. So, could it be that this sorting behavior is due to the history of my file,  and I need to create the list all over again from the scratch in case I want to see the expected result?  

This really is not a major concern, just a nice to have in the future. Also, it would be helpful to readily see how much time there still is until the alarms are fired, now one has to tap the individual alarm in order to see that info.  Could it be relocated to the place of "Everyday", with the "Everyday" being moved to the right side of that same line, immediately next to the loudspeaker icon?

**Comment by forrestguice** [comment]
> The list will always be in order of Alarm Time, but not so that the next alarm would be topmost,

It sounds like a separate bug - I think I have an idea what it looks like, but I wouldn't mind seeing a screenshot of this in action. Does closing and reopening the alarm list (re)sort the list? Or toggling between 'creation date' and 'alarm time'? Since the items are in sorted order but the first item isn't next, I'm wondering if maybe the list is left open in the background, then fails to refresh, so the next time its brought forward the order is no longer current. If so, I think closing/reopening the app, or triggering the sort should (temporarily) fix the order.  

> So, could it be that this sorting behavior is due to the history of my file

This does affect the order when sorting by 'creation date', but exporting and re-importing should fix any differences. You shouldn't need to create the alarms again. The sorting bug fixed in `0.14.7` was actually in the export - previously the file was created in the order the list was displayed (sorted however), but for 'creation date' ordering to be preserved the items must be imported in the same order they were originally created (so export needs to preserve that order).

> Also, it would be helpful to readily see how much time there still is until the alarms are fired

Seeing this info without the extra tap would be helpful. 👍 I think it might be a tight squeeze to always including the "time left" field though, so I might explore other options. Supposing the list always showed the "next" alarm on top (bug fixed), maybe automatically expanding the first item would be reasonable behavior. Or adding an option to expand/collapse the entire list at once (items always expanded). It also seems closely related to UX issues with the list itself (#611).

**Comment by Timo-Fi** [comment]
> I wouldn't mind seeing a screenshot of this in action.

Surely I'll post what ever additional info is needed. A screenshot picture of the alarm list view, in my opinion, is not very informative. A screen recorder video footage would show more, I've never used such an app but will find one, if needed.

>Does closing and reopening the alarm list (re)sort the list? 

Order of entries on the list, of each of the two views of order, is the same after a Force Stop and a device reboot, this is so on both devices.

Oh no. Please bear with me. 

This morning when I wrote the opening message, I simultaneously tested all that I wrote. And such had been the situation since yesterday evening after upgraded to 0.14.7, I was frantically testing on both devices.  But now, I did nothing else but just opened the Suntimes Alarms list, and sorting by Alarm Time is now OK on OS7. On OS11 the list is still not in expected order (again, not even Force Stop makes a change).  

Further, suddenly I now remember, this indeed is what happens on OS7, the list will eventually, after very many hours, start to be in the expected order, as if by a touch of some magic genie.  

Now, on both devices, I cleared the Alarms list, closed the app, then stated it and imported the list back. Sorting is now OK also on the OS11 device.   All is OK now.  Amazing!    And yes, I truly am absolutely sober, and have been more than 10 years :-).

Just for info: When I open the Suntimes Alarms, and then go to Sort, the most recent choice is always selected, this is Suntimes behavior. For me that is always the Alarm Time. And, when the topmost is not the next alarm that actually will trigger, I always toggle, by first doing sort by Creation Date followed by Alarm Time. I see htat the order changes to creating time order (supposedly, I can not remember in what order I have created them)  and then back to the very same Alarm Time order that it was before toggling.  Also, I always Clear (from the Recent Apps shade) apps that need not to remain "open" in background. Therefore, typically I only have the Voice Notify app on the background and that is so only on OS11,  that app does not work on OS11 if it is closed, on OS7 it can be closed.

>Supposing the list always showed the "next" alarm on top (bug fixed), maybe automatically expanding the first item would be reasonable behavior. 

Yes, expanding only the first (topmost) is a very good idea, as that is the most often sought "from now" info.  I'd like to see this info directly from a homescreen widget, but most users likely do not see it useful at all, sad to say.

>Or adding an option to expand/collapse the entire list at once (items always expanded).

YES, that would be very good indeed. When creating a lengthier list of notifications, or adding/modifying an existing, it can be very helpful to see all expanded at once. At least for me, when only one entry is expanded, it becomes difficult for me to "see"/resolve if it really is on its proper place. And, when I make some error in the definitions, the entry jumps to somewhere else on the list and it is then difficult/frustrating for me to find it again.

Thank you very much for your quick answers and I¨m really so sorry for the wasted time and effort. In the future I will avoid posting too hastily.

**Comment by forrestguice** [comment]
> the order changes to creating time order (supposedly, I can not remember in what order I have created them)

Yeah its definitely not your original order anymore. Every export done with previous versions of the app shuffled the list.

> I always Clear (from the Recent Apps shade) apps that need not to remain "open" in background

This is what I meant by closing and re-opening the app. Sorry for any confusion regarding "force stop" (although that's not a bad thing to test). It kills the running process and resets a bunch of stuff to a default state. I wanted to mention that because it also kills pending calls to `AlarmManager`, which basically breaks alarms/notifications and widget updates (until another call is made to reschedule them). 

So after doing a force stop you might notice your widgets are stale (stopped updating), and alarms/notifications fail to trigger. A device reboot will fix it, or manually toggling all the alarms disabled/enabled, or opening 'Settings->Widgets' to trigger all widgets to update (which should restore them).

> I¨m really so sorry for the wasted time and effort.

Please don't be. I appreciate the time and effort put into testing.

**Comment by Timo-Fi** [comment]
Him with the Suntimes version 0.14.10, on Android v7, v11, and v12 devices, that all have Notifications from the very same SuntimesAlarms.txt file,  with all the settings of the app the same,  and with all the settings of Android devices (likely) the same as possible.  I just took a look this again, even if it is not at all an important issue generally, it was only a small inconvenience to me when I was creating a lot of Notifications at once. So, the sort order is in timely order in all three devices, but the topmost entry on the list is not the next one that should be fired. Toggling by Sort/Creation_Date followed with Sort/Alarm_Date does not change that, nor doing ForceStop from the Android settings, nor doing device power Off/On. Interestingly, the topmost Notification is not the same Notification on these three devices, It is now 12:45 (24h notation) and I see the following:  aOS7 = 15:06, aOS11 = 13:06, and aOS12 = 19:06 (I have a Notification for each hour +/- from solar  noon, except for the nighttime).

**Comment by forrestguice** [comment]
hrm. I thought this was fixed - I'll take another look for possible causes.

Its always like this? or is only immediately after import? its still like this after a reboot? I'm wondering if maybe it starts out in the wrong order (and eventually fixes itself over time).

> but the topmost entry on the list is not the next one that should be fired

Previously, this bug could be triggered by leaving the alarm list open while notifications were triggering. Afterward that list would show the wrong order until swiped from the "recents" tray and reopened. The list should now be updating as each item triggers. It sounds like this is something a little different.

> nor doing ForceStop from the Android setting

I actually expect a 'force stop' this to kill the active alarms, with the only way to recover to either reboot, or toggling them disabled/enabled (or clear and reimport). Would it be useful to allow "reschedule all" to be manually triggered? Maybe this issue is more closely associated with import, since its frequently used as a work around to enabling/disabling all alarms.

**Comment by Timo-Fi** [comment]
> Its always like this? or is only immediately after import? its still like this after a reboot? 

With the v0.14.10, after the import of the SuntimesAlarms.txt, the list is now perfectly OK, in all my  devices. This constitutes a big change.

Reboot likely has never affected to the list in any way -or-  it is possible that it has affected in some way but has never resulted the desired situation (so that the next Notification appears at 1st position). It seems this is still so on the current version, I have not much experience with current version yet.

I seem to remember that (nearly) always the next Notification to be fired has not appeared at the 1st position (but still is always properly fired). But, I have not been inspection the list frequently (during the past nine or more months) as I'm happy with my Notifications, and Suntimes has been running smoothly. 

> wrong order until swiped 

I have not paid attention to the non-swiped items on the notification-shade, I'll keep my eye on this.  But I'm quite sure I have never any apps open ("sitting idle" on the background) unnecessarily, as I have the habit to clear them from the resent-apps list.

> I actually expect a 'force stop' this to kill the active alarms

I did the 'force stop' today, with all the three devices, to see the effect with 0.14.10.  It did not arrange the next to fire to the 1st position, that is for sure.  I seem to believe that I noticed something else but I'm not sure about it at this time, I'll come back with that if actually is so.

>Would it be useful to allow "reschedule all" to be manually triggered?

I think that this is very minor thing. As it seems now, reloading the SuntimesAlarms.txt file does the fix, that is perfectly enough for me. However, it would be nice to see, easily, at somewhere,  when the next Alarm/Notification will be fired (I would like to see days-hours-mins-secs to wait until it fires).
